providers/common/provider_util.c

```diff
@@ -164,6 +164,72 @@ ENGINE *ossl_prov_digest_engine(const PROV_DIGEST *pd)
     return pd->engine;
 }
 
+int ossl_prov_set_macctx(EVP_MAC_CTX *macctx,
+                         const OSSL_PARAM params[],
+                         const char *ciphername,
+                         const char *mdname,
+                         const char *engine,
+                         const char *properties,
+                         const unsigned char *key,
+                         size_t keylen)
+{
+    const OSSL_PARAM *p;
+    OSSL_PARAM mac_params[6], *mp = mac_params;
+
+    if (params != NULL) {
+        if (mdname == NULL) {
```

Comment:
Is it worthwhile refactoring this duplicated code into a helper function?
Something like:
``` C
int helper(const char *out_name, const char *base, const char *in_name, const OSSL_PARAM params[], OSSL_PARAM **mpp)
{
    if (base == NULL) {
        if ((p = OSSL_PARAM_locate_const(params, in_name)) != NULL) {
                if (p->data_type != OSSL_PARAM_UTF8_STRING)
                    return 0;
                base = p->data;
        }
    }
    *(*mp)++ = OSSL_PARAM_construct_utf8_string(out_name, base, 0);
    return 1;
}
```

We've also accept both utf8 strings and utf8 ptrs in places, although it isn't mandatory.

Reply:
I started to do this, but I wasn't entirely convinced it was an improvement so I backed it out again.